x-template.js

```diff
@@ -269,6 +270,1104 @@ class Forgiving {
+        `Your HTML was parsed through: \`${parsed}\`.`,
+      ].join(' ');
+    }
+    throw new Error(message);
+  }
+
+  static #validateRawString(rawString) {
+    Unforgiving.#rawUnicodeOrHexadecimal.lastIndex = 0;
+    if (Unforgiving.#rawUnicodeOrHexadecimal.test(rawString)) {
+      throw new Error([
+        `Found a unicode or hexadecimal encoding (\\x or \\u) in raw string input.`,
+        `Only valid HTML entities are supported in html.`,
+        `This is the raw string leading up to the issue \`${rawString.slice(0, Unforgiving.#rawUnicodeOrHexadecimal.lastIndex)}\`.`,
+      ].join(' '));
+    }
+  }
+
+  static #validateExit(fragment, element) {
+    if (element.value !== fragment) {
+      const tagName = element.value[Unforgiving.#localName];
+      throw new Error(`Did you forget a closing </${tagName}> at the very end of your template? To avoid unintended markup, non-void tags must explicitly be closed.`);
+    }
+  }
+
+  static #replaceHtmlEntities(originalContent) {
+    let content = originalContent;
+    Unforgiving.#htmlEntityStart.lastIndex = 0;
+    while (Unforgiving.#htmlEntityStart.test(content)) {
+      const contentIndex = Unforgiving.#htmlEntityStart.lastIndex - 2;
+      let characters = null;
+
+      Unforgiving.#entity.lastIndex = contentIndex;
+      if (!Unforgiving.#entity.test(content)) {
+        throw new Error([
+          `Seems like you have a malformed hexadecimal character reference`,
+          `(html entity). You will need to fix the reference in this content`,
+          `"${originalContent}".`,
+        ].join(' '));
+      }
+      const encoded = content.slice(contentIndex, Unforgiving.#entity.lastIndex);
+      if (Unforgiving.#hexEntity.test(encoded)) {
+        const hexadecimalDigits = encoded.slice(3, -1);
+        const number = Number(`0x${hexadecimalDigits}`);
+        try {
+          characters = String.fromCodePoint(number);
+        } catch (cause) {
+          throw new Error([
+            `Seems like you have a malformed hexadecimal character reference`,
+            `(html entity). You will need to fix the reference`,
+            `"${encoded}".`,
+          ].join(' '), { cause });
+        }
+      } else if (Unforgiving.#decEntity.test(encoded)) {
+        const decimalDigits = encoded.slice(2, -1);
+        const number = Number(decimalDigits);
+        try {
+          characters = String.fromCodePoint(number);
+        } catch (cause) {
+          throw new Error([
+            `Seems like you have a malformed decimal character reference`,
+            `(html entity). You will need to fix the reference`,
+            `"${encoded}".`,
+          ].join(' '), { cause });
+        }
+      } else {
+        // Because named character references can be written as either a decimal
+        //  or hexadecimal code points — we only support a small subset of named
+        //  entities within this parser.
+        switch (encoded) {
```

Comment:
The reason this might become frustrating is that a developer must remember *why* `&mdash;` works but `&hearts;` does not. I think I will find this surprising; having the engine THROW and EXPLAIN is potentially a good middle ground (for now &hellip...)

Reply:
Alternatively we could perhaps attempt to let the browser expand html entities using `innerHTML` carefully and selectively

Reply:
Changed, as discussed, this makes us more forgiving for invalid entities, but we decided that was ok.